design/baremetal-operator/kubebuilder-migration.md

```diff
@@ -0,0 +1,213 @@
+proof-of-concept implementation of the migration. The host API is out
+of date, and the kubebuilder code probably is as well, but the same
+basic steps can be repeated when we are ready to do the work.
+
+As much effort as possible should be made to preserve the git history
+of files such as the API definition and the controller
+implementation. Unfortunately, these significant files will also have
+to be changed beyond their location within the repository, so
+preserving all of the history may not be possible.
+
+### User Stories
+
+#### Story 1
+
+As a contributor to the baremetal-operator, I want to use familiar
+tools and workflows so I can take advantage of my experience
+contributing elsewhere in the kubernetes community to work more
+efficiently.
+
+#### Story 2
+
+As a contributor to the baremetal-operator, I want to be able to make
+API changes to the BareMetalHost and support conversion web hooks to
+change between different storage versions.
+
+#### Story 3
+
+As a contributor to the baremetal-operator, I want to be able to use
+admission web hooks to validate BareMetalHost resources.
+
+## Design Details
+
+The process for creating the proof-of-concept migration was basically:
+
+1. Move everything in the git repository out of the way.
+2. Replace the copy of operator-sdk with a copy of kubebuilder, with a
+   similar patch to support float fields.
+3. Use kubebuilder init to create a new project in the repository,
+   setting the domain to an empty string.
+4. Use kubebuilder to create an API and controller for BareMetalHost,
+   then replace the implementations with the existing ones.
+5. Restore the rest of the existing code from the pkg directory,
+   moving it to the root of the repository.
+6. Fix everything so it compiles and the tests run.
+7. Add missing targets to the Makefile (lint, sec, openapi generation,
+   etc.).
+
+### Implementation Details/Notes/Constraints
+
+Using `git mv` as often as possible will help link the new files to
+their history in any old locations.
+
+Using an empty string for the domain argument to `kubebuilder init`
+means that we can specify a group of `"metal3.io"` when creating the
+host API, and avoid ending up renaming the resource type. Future APIs
+should include a more well-scoped group ("networking.metal3.io" or
+"hardware.metal3.io") but changing the group name for the existing
+resource will make it exceptionally difficult to upgrade.
+
+### Risks and Mitigations
+
+Although operator-sdk and kubebuilder both use the same underlying
+code generation library, and are trying to achieve parity, they are
+likely to be using different versions and options from time to
+time. The new implementation will therefore introduce updated
+dependencies, and the longer we put off the work the more likely those
+dependencies are to require code changes. We will mitigate some of
+that risk by vendoring a copy of the tool, which lets us lock in a
+specific version to use.
```

Comment:
I don't think there is a need to vendor a copy of kubebuilder to lock the version. In CAPM3 we do that with the go.mod in the hack/tools folder.

Reply:
Of course modifying it to support float fields require either vendoring or maintaining our own fork of the tool and using a replace command (that would avoid huge commits in BMO). I would rather be inclined to use the second option.

Reply:
Vendoring it made it obvious how to run a specific version, so it seemed simpler. I could look at forking instead, and having the Makefile build from the fork.

Reply:
Actually I was a bit confusing things. We do not do it right now in CAPM3, we have instead an installation script that installs a specific version in a local folder : https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/hack/tools/install_kubebuilder.sh Sorry for the confusion

Reply:
OK, we could either publish built releases on GitHub and use an install script like that or we could have the scripts use an image. I'll spend a little time investigating which would be simpler.

Reply:
I've updated the text to describe the approach we discussed in the community meeting today.